crates/solver/src/driver.rs

```diff
@@ -145,13 +148,70 @@ impl Driver {
         .await
     }
 
+    async fn optimize_cost_of_unwrapping(&self, settlement: &mut Settlement) -> Result<()> {
+        let eth_for_settlement = settlement
+            .encoder
+            .unwraps()
+            .iter()
+            .find(|unwrap| unwrap.weth.address() == self.native_token)
+            .map(|unwrap| unwrap.amount)
+            .unwrap_or(U256([0; 4]));
+
+        let buffered_eth = self
+            .web3
+            .eth()
+            .balance(self.settlement_contract.address(), None)
+            .await?;
+
+        if buffered_eth >= eth_for_settlement {
+            // No need to unwrap yet, we have enough ETH buffered
+            tracing::debug!(?eth_for_settlement, "use internal buffer for unwraps");
+            settlement.encoder.drop_unwrap(self.native_token);
+            return Ok(());
+        }
+
+        let weth = contracts::WETH9::at(&self.web3, self.native_token);
+        let weth_balance = weth
```

Comment:
One complexity that I didn't think of before is, what if the solution is already planning to use part of the WETH buffer (we now allow MiP and Quasimodo to use internal buffers during their solution finding)? Then converting 60% of it to ETH may cause the settlement to fail. I'm not sure if we can somehow know the state of the buffers after the settlement was applied. 

It may require a simulation + balance query which is not trivial to implement without using [`trace_callMany`](https://openethereum.github.io/JSONRPC-trace-module#trace_callmany) (a feature of OpenEthereum, the node type we want to stop depending on) or building a [`StorageReaderExtension`](https://www.youtube.com/watch?v=4BfAEIp2_No) for our smart contract.

Alternatively, we could simulate the "optimised settlement" and discard if it fails or only perform this optimisation for our non-http-solvers (basically requiring solvers to either do all or none of the "internal buffer" optimisation)

Reply:
I assumed `SettlementEncoder::unwraps` contained all unwraps which will happen during the execution of the settlement but judging by this comment this probably only accounts for the ETH which gets paid out to the traders in the end.
Could we maybe get the needed information by analyzing `SettlementEndoder::execution_plan`? Like sum up all `Interactions` which take from our internal buffer.

Reply:
Afaik, `SettlementEncoder::unwraps` does contains all unrwaps. The problem is that we don't necessarily execute all swaps the are needed for the settlement to "perfectly clear" (sample settlement: https://etherscan.io/tx/0xec7fcc9c59e3b56d29ea7feb3f79afc2fb1c0c73e7c0049075be1666a64f2ced, see how we do a bunch of trades without any external swap - the imbalance is paid from "internal buffers")

Here is an example where this new logic would fail: Let's say we have 10 WETH (0 ETH) in the contract and there is a trade selling 20k USDC for 5 WETH together with a trade selling some shitcoin (which is not "whitelisted" for internal buffer trading) for 1 ETH. 
Since recently, our http solvers receive information about the buffers as part of the auction instance and may - in order to be more gas efficient - decide to use less interactions (e.g. skip the $20k -> 5 WETH swap) and pay those from internal buffers.

In this case the logic here would decide to unwrap 6 WETH to ETH, however this wouldn't leave enough WETH to pay for the $20k -> 5 WETH trade.

Judging the expected in/out amounts from interactions would require adding the in/out token + amounts to the `Interaction` trait - something we have spoken about doing in the past (it might also be relevant for a smarter solutions submission strategy that @sunce86 is working on). I'm not sure how big of a refactoring this would be and if it would be simpler for now to simulate the optimised solution (maybe only necessary if we add an unwrap) and discard it if it's no longer valid.